client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java

```diff
@@ -51,21 +51,20 @@ public T execute() {
             transport = getTransport(retryCount, failedServers);
             return executeOperation(transport);
          } catch (TransportException te) {
-            if (failedServers == null) {
-               failedServers = new HashSet<SocketAddress>();
-            }
-            failedServers.add(te.getServerAddress());
+            SocketAddress address = te.getServerAddress();
+            failedServers = updateFailedServers(address, failedServers);
             // Invalidate transport since this exception means that this
             // instance is no longer usable and should be destroyed.
-            if (transport != null) {
-               if (log.isTraceEnabled())
-                  log.tracef("Invalidating transport %s as a result of transport exception", transport);
-
-               transportFactory.invalidateTransport(
-                     te.getServerAddress(), transport);
-            }
+            invalidateTransport(transport, address);
             retryCount = logTransportErrorAndThrowExceptionIfNeeded(retryCount, te);
-         } catch (RemoteNodeSuspectException | RemoteIllegalLifecycleStateException e) {
+         } catch (RemoteIllegalLifecycleStateException e) {
+            SocketAddress address = e.getServerAddress();
+            failedServers = updateFailedServers(address, failedServers);
+            // Invalidate transport since this exception means that this
+            // instance is no longer usable and should be destroyed.
+            invalidateTransport(transport, address);
+            retryCount = logTransportErrorAndThrowExceptionIfNeeded(retryCount, e);
```

Comment:
It could be done, but the fact that the current design updates 2 local variables make it a bit more awkward to do, and did not want to refactor that at this point.